diff --git a/molopt/base.py b/molopt/base.py
index a9ad560..680daa7 100644
--- a/molopt/base.py
+++ b/molopt/base.py
@@ -8,6 +8,9 @@ from rdkit.Chem import Draw
 import tdc
 from tdc.generation import MolGen
 from .utils.chem import *
+import joblib
+from joblib import Parallel, delayed
+
 
 
 class Objdict(dict):
@@ -48,7 +51,7 @@ def top_auc(buffer, top_n, finish, freq_log, max_oracle_calls):
 
 
 class Oracle:
-    def __init__(self, max_oracle_calls=10000, freq_log=100, output_dir='results', mol_buffer={}):
+    def __init__(self, max_oracle_calls=10000, freq_log=100, output_dir='results', mol_buffer=None):
         self.name = None
         self.evaluator = None
         self.task_label = None
@@ -57,7 +60,7 @@ class Oracle:
         self.freq_log = freq_log 
         self.output_dir = output_dir 
 
-        self.mol_buffer = mol_buffer
+        self.mol_buffer = {} if mol_buffer is None else mol_buffer
         self.sa_scorer = tdc.Oracle(name = 'SA')
         self.diversity_evaluator = tdc.Evaluator(name = 'Diversity')
         self.last_log = 0
@@ -206,7 +209,7 @@ class BaseOptimizer:
     def __init__(self, smi_file=None, n_jobs=-1, max_oracle_calls=10000, freq_log=100, output_dir = 'results', log_results=True):
         self.model_name = "Default"
         self.n_jobs = n_jobs
-        # self.pool = joblib.Parallel(n_jobs=self.n_jobs)
+        self.pool = joblib.Parallel(n_jobs=self.n_jobs)
         self.smi_file = smi_file
         self.max_oracle_calls = max_oracle_calls 
         self.freq_log = freq_log 
@@ -225,9 +228,9 @@ class BaseOptimizer:
         self.diversity_evaluator = tdc.Evaluator(name = 'Diversity')
         self.filter = tdc.chem_utils.oracle.filter.MolFilter(filters = ['PAINS', 'SureChEMBL', 'Glaxo'], property_filters_flag = False)
 
-    # def load_smiles_from_file(self, file_name):
-    #     with open(file_name) as f:
-    #         return self.pool(delayed(canonicalize)(s.strip()) for s in f)
+    def load_smiles_from_file(self, file_name):
+        with open(file_name) as f:
+            return self.pool(delayed(canonicalize)(s.strip()) for s in f)
             
     def sanitize(self, mol_list):
         new_mol_list = []
@@ -354,4 +357,3 @@ class BaseOptimizer:
         for seed in seeds:
             self.optimize(oracle, config, seed, project)
             self.reset()
-
diff --git a/molopt/mimosa/online_train.py b/molopt/mimosa/online_train.py
index 3fe9074..12d2c0a 100644
--- a/molopt/mimosa/online_train.py
+++ b/molopt/mimosa/online_train.py
@@ -7,7 +7,7 @@ from tqdm import tqdm
 from random import shuffle 
 from module import GCN 
 from chemutils import smiles2feature  
-from utils import Molecule_Dataset 
+from .utils import Molecule_Dataset
 device = 'cpu'
 
 def train_gnn(data, gnn, epoch=5): 
diff --git a/molopt/selfies_ga/discriminator.py b/molopt/selfies_ga/discriminator.py
index 52b4596..86ead5a 100644
--- a/molopt/selfies_ga/discriminator.py
+++ b/molopt/selfies_ga/discriminator.py
@@ -31,7 +31,7 @@ def create_discriminator(one_hot_len, n_hidden, device):
     """
     n_hidden.insert(0, one_hot_len)
 
-    net = Net(n_feature=one_hot_len, h_sizes=n_hidden, n_output=1).to(device)
+    net = Net(n_feature=one_hot_len, h_sizes=n_hidden, n_output=1).float().to(device)
     optimizer = torch.optim.Adam(net.parameters(), lr=0.001, weight_decay=1e-4)
     loss_func = torch.nn.BCELoss()
     
@@ -57,7 +57,7 @@ def obtain_initial_discriminator(disc_enc_type, disc_layers, max_molecules_len,
     if disc_enc_type == 'smiles' or disc_enc_type == 'selfies':
         alphabet = smiles_alphabet(disc_enc_type)
         one_hot_len = len(alphabet) * max_molecules_len
-        discriminator, d_optimizer, d_loss_func = create_discriminator(one_hot_len, disc_layers, 0.0, device) 
+        discriminator, d_optimizer, d_loss_func = create_discriminator(one_hot_len, disc_layers, device)
         return discriminator, d_optimizer, d_loss_func
     elif disc_enc_type == 'properties_rdkit':
         discriminator, d_optimizer, d_loss_func = create_discriminator(51, disc_layers, device)  
@@ -155,5 +155,3 @@ def check_discr_improvement(g_data, data_y, generation_index, ga_discr_lookback,
 
     
 
-
-
